pkg/engine/k8smanifest.go

```diff
@@ -0,0 +1,401 @@
+	verifiedCount := 0
+	attestorSet = expandStaticKeys(attestorSet)
+	requiredCount := getRequiredCount(attestorSet)
+	errorList := []error{}
+	verifiedMessageList := []string{}
+	failedMessageList := []string{}
+
+	for i, a := range attestorSet.Entries {
+		var entryError error
+		var verified bool
+		var reason string
+		attestorPath := fmt.Sprintf("%s.entries[%d]", path, i)
+		if a.Attestor != nil {
+			nestedAttestorSet, err := kyvernov1.AttestorSetUnmarshal(a.Attestor)
+			if err != nil {
+				entryError = errors.Wrapf(err, "failed to unmarshal nested attestor %s", attestorPath)
+			} else {
+				attestorPath += ".attestor"
+				verified, reason, err = verifyManifestAttestorSet(resource, *nestedAttestorSet, vo, attestorPath, uid, logger)
+				if err != nil {
+					entryError = errors.Wrapf(err, "failed to verify signature; %s", attestorPath)
+				}
+			}
+		} else {
+			verified, reason, entryError = k8sVerifyResource(resource, a, vo, attestorPath, uid, i, logger)
+		}
+
+		if entryError != nil {
+			errorList = append(errorList, entryError)
+		} else if verified {
+			// verification success.
+			verifiedCount++
+			verifiedMessageList = append(verifiedMessageList, reason)
+		} else {
+			failedMessageList = append(failedMessageList, reason)
+		}
+
+		if verifiedCount >= requiredCount {
+			logger.V(2).Info("manifest verification succeeded", "verifiedCount", verifiedCount, "requiredCount", requiredCount)
+			reason := fmt.Sprintf("manifest verification succeeded; verifiedCount %d; requiredCount %d; message %s",
+				verifiedCount, requiredCount, strings.Join(verifiedMessageList, ","))
+			return true, reason, nil
+		}
+	}
+
+	if len(errorList) != 0 {
+		err := multierr.Combine(errorList...)
+		logger.V(2).Info("manifest verification failed", "verifiedCount", verifiedCount, "requiredCount",
+			requiredCount, "errors", errorList)
+		return false, "", err
+	}
+	reason := fmt.Sprintf("manifest verification failed; verifiedCount %d; requiredCount %d; message %s",
+		verifiedCount, requiredCount, strings.Join(failedMessageList, ","))
+	logger.V(2).Info("manifest verification failed", "verifiedCount", verifiedCount, "requiredCount",
+		requiredCount, "reason", failedMessageList)
+	return false, reason, nil
+}
+
+func k8sVerifyResource(resource unstructured.Unstructured, a kyvernov1.Attestor, vo *k8smanifest.VerifyResourceOption, attestorPath, uid string, i int, logger logr.Logger) (bool, string, error) {
+	// check annotations
+	if a.Annotations != nil {
+		mnfstAnnotations := resource.GetAnnotations()
+		err := checkManifestAnnotations(mnfstAnnotations, a.Annotations)
+		if err != nil {
+			return false, "", err
+		}
+	}
+
+	// build verify option
+	vo, subPath, envVariables, err := buildVerifyResourceOptionsAndPath(a, vo, uid, i)
+	// unset env variables after verification
+	defer cleanEnvVariables(envVariables)
```

Comment:
Even though we clean Environment variables after each invocation, there may be a chance Kyverno processes multiple requests in parallel, and a shared environment will cause an issue.